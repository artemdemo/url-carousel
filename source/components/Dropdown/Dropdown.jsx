import React from 'react';
import classnames from 'classnames';

import './Dropdown.less';

const Dropdown = (props) => {
    const { onChange, value, placeholder, disabled, error, list = [] } = props;

    const itemsList = (() => {
        if (placeholder) {
            return [placeholder].concat(list);
        }
        return list;
    })();

    const dropdownValue = (() => {
        if ((value === '' || value === undefined) && placeholder) {
            return '-1';
        }
        return value;
    })();

    const renderErrorText = (error) => {
        if (typeof error === 'string' && error !== '') {
            return (
                <div className='dropdown-error'>
                    {error}
                </div>
            );
        }
        return null;
    };

    const containerClass = classnames({
        'has-error': Boolean(error),
    });

    return (
        <div className={containerClass}>
            <select
                className='form-control'
                value={dropdownValue}
                onChange={e => onChange && onChange(Number(e.target.value))}
                disabled={disabled}>
                {itemsList.map((item, index) => {
                    return (
                        <option value={placeholder ? index - 1 : index}
                                disabled={placeholder && index - 1 === -1}
                                key={`dropdown__item-${index}`}>{item}</option>
                    );
                })}
            </select>
            {renderErrorText(error)}
        </div>
    );
};

Dropdown.propTypes = {
    onChange: React.PropTypes.func,
    className: React.PropTypes.string,
    list: React.PropTypes.arrayOf(React.PropTypes.string),
    value: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    disabled: React.PropTypes.bool,
    error: React.PropTypes.oneOfType([
        React.PropTypes.bool,
        React.PropTypes.string,
    ]),
};

export default Dropdown;
