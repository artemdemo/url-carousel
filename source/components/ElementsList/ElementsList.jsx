import React from 'react';
import classnames from 'classnames';

import './ElementsList.less';

/**
 * <ElementsList /> is used to add margin between elements
 */
const ElementsList = (props) => {
    let children;
    let counter = 0;
    const childAmount = props.children.length;
    if (childAmount > 1) {
        children = React.Children.map(
            props.children,
            (child) => {
                if (child) {
                    const newProps = {
                        className: classnames(child.props.className, {
                            'elements-list-item': true,
                            'elements-list-item_first': counter === 0,
                            'elements-list-item_last': counter === childAmount - 1,
                        }),
                    };
                    counter++;
                    return React.cloneElement(child, newProps);
                }
                return null;
            });
    } else {
        children = props.children;
    }
    const { inline, nowrap } = props;
    const wrapClass = classnames({
        'elements-list': true,
        'elements-list_inline': inline,
        'elements-list_nowrap': nowrap,
    });
    return (
        <div className={wrapClass}>
            {children}
        </div>
    );
};

ElementsList.propTypes = {
    inline: React.PropTypes.bool,
    nowrap: React.PropTypes.bool,
};

export default ElementsList;
