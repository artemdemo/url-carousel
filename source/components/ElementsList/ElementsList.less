.elements-list {
    display: block;
}
.elements-list_inline {
    display: inline-block;
}
.elements-list_nowrap {
    white-space: nowrap;
}

.elements-list-item {
    margin-right: 7px;
}

.elements-list-item_first {}

.elements-list-item_last {
    margin-right: 0;
}
