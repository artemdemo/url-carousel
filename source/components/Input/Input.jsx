import React from 'react';
import classnames from 'classnames';

import './Input.less';

const Input = (props) => {
    const { type = 'text', value, placeholder, disabled, error, onChange } = props;

    const renderErrorText = () => {
        if (typeof error === 'string' && error !== '') {
            return (
                <div className='input-error text-danger'>
                    {error}
                </div>
            );
        }
        return null;
    };

    const containerClass = classnames({
        'has-error': Boolean(error),
    });

    return (
        <div className={containerClass}>
            <input
                className='form-control'
                type={type}
                value={value}
                placeholder={placeholder}
                disabled={disabled}
                onChange={e => onChange && onChange(e.target.value)} />
            {renderErrorText()}
        </div>
    );
};

Input.propTypes = {
    type: React.PropTypes.string,
    value: React.PropTypes.string,
    placeholder: React.PropTypes.string,
    onChange: React.PropTypes.func,
    disabled: React.PropTypes.bool,
    error: React.PropTypes.oneOfType([
        React.PropTypes.bool,
        React.PropTypes.string,
    ]),
};

export default Input;
