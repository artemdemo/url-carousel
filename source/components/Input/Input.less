.input-error {
    padding-left: 5px;
}
