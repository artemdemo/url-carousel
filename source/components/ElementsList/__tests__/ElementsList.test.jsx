import React from 'react';
import renderer from 'react-test-renderer';
import Input from '../../Input/Input';
import Dropdown from '../../Dropdown/Dropdown';
import ElementsList from '../ElementsList';

describe('<ElementsList>', () => {
    it('Render <ElementsList>', () => {
        const tree = renderer.create(
            <ElementsList>
                <Input />
                <span>Some dropdown:</span>
                <Dropdown list={['First item in dropdown', 'Second item in dropdown']} />
            </ElementsList>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Render <ElementsList inline>', () => {
        const tree = renderer.create(
            <ElementsList inline>
                <div>First</div>
                <div>First</div>
            </ElementsList>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Render <ElementsList> with null', () => {
        const tree = renderer.create(
            <ElementsList inline>
                {null}
                <div>First</div>
            </ElementsList>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
