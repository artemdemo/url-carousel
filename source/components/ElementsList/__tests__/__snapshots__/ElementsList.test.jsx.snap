// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ElementsList> Render <ElementsList inline> 1`] = `
<div
  className="elements-list elements-list_inline"
>
  <div
    className="elements-list-item elements-list-item_first"
  >
    First
  </div>
  <div
    className="elements-list-item elements-list-item_last"
  >
    First
  </div>
</div>
`;

exports[`<ElementsList> Render <ElementsList> 1`] = `
<div
  className="elements-list"
>
  <div
    className="input-container"
  >
    <input
      className="form-control"
      disabled={undefined}
      onChange={[Function]}
      placeholder={undefined}
      type="text"
      value={undefined}
    />
  </div>
  <span
    className="elements-list-item"
  >
    Some dropdown:
  </span>
  <div
    className="elements-list-item elements-list-item_last"
  >
    <select
      className="form-control"
      disabled={undefined}
      onChange={[Function]}
      value={undefined}
    >
      <option
        disabled={undefined}
        value={0}
      >
        First item in dropdown
      </option>
      <option
        disabled={undefined}
        value={1}
      >
        Second item in dropdown
      </option>
    </select>
  </div>
</div>
`;

exports[`<ElementsList> Render <ElementsList> with null 1`] = `
<div
  className="elements-list elements-list_inline"
>
  <div
    className="elements-list-item elements-list-item_first"
  >
    First
  </div>
</div>
`;
